Diabeties Dataset Cleaning and Exploration

| Feature Name               | Type    | Description | % Missing |
|----------------------------|---------|-------------|-----------|
| Encounter ID               | Numeric | Unique identifier of an encounter | 0% |
| Patient number             | Numeric | Unique identifier of a patient | 0% |
| Race                       | Nominal | Values: Caucasian, Asian, African American, Hispanic, and other | 2% |
| Gender                     | Nominal | Values: male, female, and unknown/invalid | 0% |
| Age                        | Nominal | Grouped in 10-year intervals: [0, 10), [10, 20), ..., [90, 100) | 0% |
| Weight                     | Numeric | Weight in pounds | 97% |
| Admission type             | Nominal | Integer identifier corresponding to 9 distinct values, e.g., emergency, urgent, elective, newborn, and not available | 0% |
| Discharge disposition      | Nominal | Integer identifier corresponding to 29 distinct values, e.g., discharged to home, expired, and not available | 0% |
| Admission source           | Nominal | Integer identifier corresponding to 21 distinct values, e.g., physician referral, emergency room, and transfer from a hospital | 0% |
| Time in hospital           | Numeric | Integer number of days between admission and discharge | 0% |
| Payer code                 | Nominal | Integer identifier corresponding to 23 distinct values, e.g., Blue Cross/Blue Shield, Medicare, and self-pay | 52% |
| Medical specialty          | Nominal | Integer identifier of the admitting physician's specialty (84 distinct values), e.g., cardiology, internal medicine, family/general practice, and surgeon | 53% |
| Number of lab procedures   | Numeric | Number of lab tests performed during the encounter | 0% |
| Number of procedures       | Numeric | Number of procedures (other than lab tests) performed during the encounter | 0% |
| Number of medications      | Numeric | Number of distinct generic names administered during the encounter | 0% |
| Number of outpatient visits | Numeric | Number of outpatient visits of the patient in the year preceding the encounter | 0% |
| Number of emergency visits | Numeric | Number of emergency visits of the patient in the year preceding the encounter | 0% |
| Number of inpatient visits | Numeric | Number of inpatient visits of the patient in the year preceding the encounter | 0% |
| Diagnosis 1               | Nominal | Primary diagnosis (coded as first three digits of ICD9); 848 distinct values | 0% |
| Diagnosis 2               | Nominal | Secondary diagnosis (coded as first three digits of ICD9); 923 distinct values | 0% |
| Diagnosis 3               | Nominal | Additional secondary diagnosis (coded as first three digits of ICD9); 954 distinct values | 1% |
| Number of diagnoses       | Numeric | Number of diagnoses entered to the system | 0% |
| Glucose serum test result | Nominal | Indicates the range of the result or if the test was not taken. Values: ">200", ">300", "normal", "none" | 0% |
| A1c test result           | Nominal | Indicates the range of the result or if the test was not taken. Values: ">8", ">7", "normal", "none" | 0% |
| Change of medications     | Nominal | Indicates if there was a change in diabetic medications. Values: "change" and "no change" | 0% |
| Diabetes medications      | Nominal | Indicates if any diabetic medication was prescribed. Values: "yes" and "no" | 0% |
| 24 features for medications | Nominal | Each feature represents a drug (e.g., metformin, insulin) and indicates whether it was prescribed or dosage changed. Values: "up", "down", "steady", "no" | 0% |
| Readmitted                | Nominal | Days to inpatient readmission. Values: "<30", ">30", "No" | 0% |


In [2]:
import pandas as pd
import numpy as np

In [47]:
data = pd.read_csv('diabetic_data.csv')

In [48]:
data

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)      ?   
1             149190     55629189        Caucasian  Female  [10-20)      ?   
2              64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3             500364     82442376        Caucasian    Male  [30-40)      ?   
4              16680     42519267        Caucasian    Male  [40-50)      ?   
...              ...          ...              ...     ...      ...    ...   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)      ?   
101762     443847782     74694222  AfricanAmerican  Female  [80-90)      ?   
101763     443854148     41088789        Caucasian    Male  [70-80)      ?   
101764     443857166     31693671        Caucasian  Female  [80-90)      ?   
101765     443867222    175429310        Caucasian    Male  [70-80)      ?   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
2                       1                         1                    7   
3                       1                         1                    7   
4                       1                         1                    7   
...                   ...                       ...                  ...   
101761                  1                         3                    7   
101762                  1                         4                    5   
101763                  1                         1                    7   
101764                  2                         3                    7   
101765                  1                         1                    7   

        time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                      1  ...          No      No                   No   
1                      3  ...          No      Up                   No   
2                      2  ...          No      No                   No   
3                      2  ...          No      Up                   No   
4                      1  ...          No  Steady                   No   
...                  ...  ...         ...     ...                  ...   
101761                 3  ...          No    Down                   No   
101762                 5  ...          No  Steady                   No   
101763                 1  ...          No    Down                   No   
101764                10  ...          No      Up                   No   
101765                 6  ...          No      No                   No   

        glipizide-metformin  glimepiride-pioglitazone  \
0                        No                        No   
1                        No                        No   
2                        No                        No   
3                        No                        No   
4                        No                        No   
...                     ...                       ...   
101761                   No                        No   
101762                   No                        No   
101763                   No                        No   
101764                   No                        No   
101765                   No                        No   

        metformin-rosiglitazone  metformin-pioglitazone  change diabetesMed  \
0                            No                      No      No          No   
1                            No                      No      Ch         Yes   
2                            No                      No      No         Yes   
3                            No                      No      Ch         Yes   
4                            No                      No      Ch         Yes   
...                         ...                     ...     ...         ...   
101761                       No

In [49]:
data.columns

Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted'],
      dtype='object')

In [50]:
data['weight'].unique()

array(['?', '[75-100)', '[50-75)', '[0-25)', '[100-125)', '[25-50)',
       '[125-150)', '[175-200)', '[150-175)', '>200'], dtype=object)

In [51]:
data.shape

(101766, 50)

In [52]:
data[data['weight'] == '?']

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)      ?   
1             149190     55629189        Caucasian  Female  [10-20)      ?   
2              64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3             500364     82442376        Caucasian    Male  [30-40)      ?   
4              16680     42519267        Caucasian    Male  [40-50)      ?   
...              ...          ...              ...     ...      ...    ...   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)      ?   
101762     443847782     74694222  AfricanAmerican  Female  [80-90)      ?   
101763     443854148     41088789        Caucasian    Male  [70-80)      ?   
101764     443857166     31693671        Caucasian  Female  [80-90)      ?   
101765     443867222    175429310        Caucasian    Male  [70-80)      ?   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
2                       1                         1                    7   
3                       1                         1                    7   
4                       1                         1                    7   
...                   ...                       ...                  ...   
101761                  1                         3                    7   
101762                  1                         4                    5   
101763                  1                         1                    7   
101764                  2                         3                    7   
101765                  1                         1                    7   

        time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                      1  ...          No      No                   No   
1                      3  ...          No      Up                   No   
2                      2  ...          No      No                   No   
3                      2  ...          No      Up                   No   
4                      1  ...          No  Steady                   No   
...                  ...  ...         ...     ...                  ...   
101761                 3  ...          No    Down                   No   
101762                 5  ...          No  Steady                   No   
101763                 1  ...          No    Down                   No   
101764                10  ...          No      Up                   No   
101765                 6  ...          No      No                   No   

        glipizide-metformin  glimepiride-pioglitazone  \
0                        No                        No   
1                        No                        No   
2                        No                        No   
3                        No                        No   
4                        No                        No   
...                     ...                       ...   
101761                   No                        No   
101762                   No                        No   
101763                   No                        No   
101764                   No                        No   
101765                   No                        No   

        metformin-rosiglitazone  metformin-pioglitazone  change diabetesMed  \
0                            No                      No      No          No   
1                            No                      No      Ch         Yes   
2                            No                      No      No         Yes   
3                            No                      No      Ch         Yes   
4                            No                      No      Ch         Yes   
...                         ...                     ...     ...         ...   
101761                       No

98569 weights are missing out of 101766. Do not use weight as a factor for prediction model

In [53]:
data.isna().sum().sort_values(ascending=False)

max_glu_serum               96420
A1Cresult                   84748
race                            0
gender                          0
age                             0
weight                          0
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
encounter_id                    0
patient_nbr                     0
number_inpatient                0
number_emergency                0
diag_1                          0
diag_2                          0
number_diagnoses                0
diag_3                          0
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

max_glu_serum and a1cresult should not be used either, lots of missing values

Filter dataset to only include diabetes diagnoses

In [54]:
def is_numeric(s):
    try:
        float(s)
        return True
    except ValueError:
        return False

In [55]:
data['diabetes_1'] = data['diag_1'].apply(lambda x: float(x) if is_numeric(x) else np.nan)
data['diabetes_2'] = data['diag_2'].apply(lambda x: float(x) if is_numeric(x) else np.nan)
data['diabetes_3'] = data['diag_3'].apply(lambda x: float(x) if is_numeric(x) else np.nan)

In [63]:
data = data[((data['diabetes_1'] > 250) & (data['diabetes_1'] < 251)) 
          | ((data['diabetes_2'] > 250) & (data['diabetes_2'] < 251)) 
          | ((data['diabetes_3'] > 250) & (data['diabetes_3'] < 251))]


Replace weight '?' with np.nan

In [59]:
data['weight'].replace('?', np.nan, inplace=True)

C:\Users\lamet\AppData\Local\Temp\ipykernel_2196\742919771.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['weight'].replace('?', np.nan, inplace=True)


In [64]:
data

encounter_id  patient_nbr             race  gender      age weight  \
0            2278392      8222157        Caucasian  Female   [0-10)    NaN   
1             149190     55629189        Caucasian  Female  [10-20)    NaN   
3             500364     82442376        Caucasian    Male  [30-40)    NaN   
10             28236     89869032  AfricanAmerican  Female  [40-50)    NaN   
12             40926     85504905        Caucasian  Female  [40-50)    NaN   
...              ...          ...              ...     ...      ...    ...   
101736     443739152     90751788        Caucasian  Female  [60-70)    NaN   
101742     443787512     52419276        Caucasian    Male  [70-80)    NaN   
101753     443841992    184875899            Other    Male  [40-50)    NaN   
101754     443842016    183087545        Caucasian  Female  [70-80)    NaN   
101761     443847548    100162476  AfricanAmerican    Male  [70-80)    NaN   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
3                       1                         1                    7   
10                      1                         1                    7   
12                      1                         3                    7   
...                   ...                       ...                  ...   
101736                  1                         3                    7   
101742                  2                         6                    2   
101753                  1                         1                    7   
101754                  1                         1                    7   
101761                  1                         3                    7   

        time_in_hospital  ... glipizide-metformin glimepiride-pioglitazone  \
0                      1  ...                  No                       No   
1                      3  ...                  No                       No   
3                      2  ...                  No                       No   
10                     9  ...                  No                       No   
12                     7  ...                  No                       No   
...                  ...  ...                 ...                      ...   
101736                 8  ...                  No                       No   
101742                 4  ...                  No                       No   
101753                13  ...                  No                       No   
101754                 9  ...                  No                       No   
101761                 3  ...                  No                       No   

        metformin-rosiglitazone  metformin-pioglitazone  change  diabetesMed  \
0                            No                      No      No           No   
1                            No                      No      Ch          Yes   
3                            No                      No      Ch          Yes   
10                           No                      No      No          Yes   
12                           No                      No      Ch          Yes   
...                         ...                     ...     ...          ...   
101736                       No                      No      No          Yes   
101742                       No                      No      Ch          Yes   
101753                       No                      No      Ch          Yes   
101754                       No                      No      Ch          Yes   
101761                       No                      No      Ch          Yes   

        readmitted  diabetes_1 diabetes_2 diabetes_3  
0               NO      250.83        NaN        NaN  
1              >30      276.00     250.01     255.00  
3               NO        8.00     250.43     403.00  
10             >30      250.70     403.0

Handling missing values

In [65]:
data.isna().sum().sort_values(ascending=False)

weight                      19798
max_glu_serum               18862
A1Cresult                   14659
diabetes_3                   1621
diabetes_2                    636
diabetes_1                    195
age                             0
discharge_disposition_id        0
admission_type_id               0
time_in_hospital                0
payer_code                      0
medical_specialty               0
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
admission_source_id             0
patient_nbr                     0
race                            0
gender                          0
encounter_id                    0
diag_3                          0
diag_2                          0
diag_1                          0
number_inpatient                0
number_emergency                0
repaglinide                     0
number_diagnoses                0
metformin                       0
acetohexamide 

We leave diabeties 1, 2, 3 na values as is because we should not impute a diagnoses. We are only trying to focus on the diagnoses of diabeties

Look for '?' and replace with na values

In [67]:
data.replace('?', np.nan, inplace=True)

C:\Users\lamet\AppData\Local\Temp\ipykernel_2196\2483408281.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.replace('?', np.nan, inplace=True)


Remove irrelevant columns

'Encounter_id': We will use index as unique identifier

'Diagnosis 1,2,3': We filtered our dataset to if a paitent was diagnosed with diabeties either or through 1,2,3


In [82]:
data = data.drop(['encounter_id', 'diag_1', 'diag_2', 'diag_3'], axis=1)

In [83]:
data

patient_nbr             race  gender      age weight  \
0           8222157        Caucasian  Female   [0-10)    NaN   
1          55629189        Caucasian  Female  [10-20)    NaN   
3          82442376        Caucasian    Male  [30-40)    NaN   
10         89869032  AfricanAmerican  Female  [40-50)    NaN   
12         85504905        Caucasian  Female  [40-50)    NaN   
...             ...              ...     ...      ...    ...   
101736     90751788        Caucasian  Female  [60-70)    NaN   
101742     52419276        Caucasian    Male  [70-80)    NaN   
101753    184875899            Other    Male  [40-50)    NaN   
101754    183087545        Caucasian  Female  [70-80)    NaN   
101761    100162476  AfricanAmerican    Male  [70-80)    NaN   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
3                       1                         1                    7   
10                      1                         1                    7   
12                      1                         3                    7   
...                   ...                       ...                  ...   
101736                  1                         3                    7   
101742                  2                         6                    2   
101753                  1                         1                    7   
101754                  1                         1                    7   
101761                  1                         3                    7   

        time_in_hospital payer_code  ... glipizide-metformin  \
0                      1        NaN  ...                  No   
1                      3        NaN  ...                  No   
3                      2        NaN  ...                  No   
10                     9        NaN  ...                  No   
12                     7        NaN  ...                  No   
...                  ...        ...  ...                 ...   
101736                 8         HM  ...                  No   
101742                 4         MC  ...                  No   
101753                13        NaN  ...                  No   
101754                 9        NaN  ...                  No   
101761                 3         MC  ...                  No   

        glimepiride-pioglitazone  metformin-rosiglitazone  \
0                             No                       No   
1                             No                       No   
3                             No                       No   
10                            No                       No   
12                            No                       No   
...                          ...                      ...   
101736                        No                       No   
101742                        No                       No   
101753                        No                       No   
101754                        No                       No   
101761                        No                       No   

        metformin-pioglitazone  change  diabetesMed  readmitted  diabetes_1  \
0                           No      No           No          NO      250.83   
1                           No      Ch          Yes         >30      276.00   
3                           No      Ch          Yes          NO        8.00   
10                          No      No          Yes         >30      250.70   
12                          No      Ch          Yes         <30      428.00   
...                        ...     ...          ...         ...         ...   
101736                      No      No          Yes         >30      590.00   
101742                      No      Ch          Yes          NO      250.80   
101753                      No      Ch          Yes          NO      250.80   
101754                      No    

Make sure data types are correct

Age and weight are binned, so strings instead of ints

Diabeties 1, 2, 3 are IC9 codes which all mean this patient was diagnosed with diabetes

Medication columns include values 'Yes', 'No', 'Steady', 'Up', 'Down'. We did not convert these columns to 1,0 because there are multiple values, not just yes, no

In [103]:
data.columns

Index(['patient_nbr', 'race', 'gender', 'age', 'weight', 'admission_type_id',
       'discharge_disposition_id', 'admission_source_id', 'time_in_hospital',
       'payer_code', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'number_diagnoses',
       'max_glu_serum', 'A1Cresult', 'metformin', 'repaglinide', 'nateglinide',
       'chlorpropamide', 'glimepiride', 'acetohexamide', 'glipizide',
       'glyburide', 'tolbutamide', 'pioglitazone', 'rosiglitazone', 'acarbose',
       'miglitol', 'troglitazone', 'tolazamide', 'examide', 'citoglipton',
       'insulin', 'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'diabetes_1', 'diabetes_2', 'diabetes_3'],
      dtype='object')

In [104]:
data['insulin'].value_counts()

insulin
Steady    6666
No        5881
Down      4031
Up        3588
Name: count, dtype: int64

In [107]:
data['change'].value_counts()

change
Ch    11224
No     8942
Name: count, dtype: int64

Final Dataset

In [108]:
data

patient_nbr             race  gender      age weight  \
0           8222157        Caucasian  Female   [0-10)    NaN   
1          55629189        Caucasian  Female  [10-20)    NaN   
3          82442376        Caucasian    Male  [30-40)    NaN   
10         89869032  AfricanAmerican  Female  [40-50)    NaN   
12         85504905        Caucasian  Female  [40-50)    NaN   
...             ...              ...     ...      ...    ...   
101736     90751788        Caucasian  Female  [60-70)    NaN   
101742     52419276        Caucasian    Male  [70-80)    NaN   
101753    184875899            Other    Male  [40-50)    NaN   
101754    183087545        Caucasian  Female  [70-80)    NaN   
101761    100162476  AfricanAmerican    Male  [70-80)    NaN   

        admission_type_id  discharge_disposition_id  admission_source_id  \
0                       6                        25                    1   
1                       1                         1                    7   
3                       1                         1                    7   
10                      1                         1                    7   
12                      1                         3                    7   
...                   ...                       ...                  ...   
101736                  1                         3                    7   
101742                  2                         6                    2   
101753                  1                         1                    7   
101754                  1                         1                    7   
101761                  1                         3                    7   

        time_in_hospital payer_code  ... glipizide-metformin  \
0                      1        NaN  ...                  No   
1                      3        NaN  ...                  No   
3                      2        NaN  ...                  No   
10                     9        NaN  ...                  No   
12                     7        NaN  ...                  No   
...                  ...        ...  ...                 ...   
101736                 8         HM  ...                  No   
101742                 4         MC  ...                  No   
101753                13        NaN  ...                  No   
101754                 9        NaN  ...                  No   
101761                 3         MC  ...                  No   

        glimepiride-pioglitazone  metformin-rosiglitazone  \
0                             No                       No   
1                             No                       No   
3                             No                       No   
10                            No                       No   
12                            No                       No   
...                          ...                      ...   
101736                        No                       No   
101742                        No                       No   
101753                        No                       No   
101754                        No                       No   
101761                        No                       No   

        metformin-pioglitazone  change  diabetesMed  readmitted  diabetes_1  \
0                           No      No           No          NO      250.83   
1                           No      Ch          Yes         >30      276.00   
3                           No      Ch          Yes          NO        8.00   
10                          No      No          Yes         >30      250.70   
12                          No      Ch          Yes         <30      428.00   
...                        ...     ...          ...         ...         ...   
101736                      No      No          Yes         >30      590.00   
101742                      No      Ch          Yes          NO      250.80   
101753                      No      Ch          Yes          NO      250.80   
101754                      No    